A1.md

```diff
@@ -0,0 +1,172 @@
+  can dynamically choose whether to use HTTP CONNECT and to which proxy
+  based on the server name and/or server address.
+  - In C-core, this will be implemented as follows: In the subchannel,
+    before we connect to the target address, we will call a new *proxy
+    mapper* hook, which will allow selectively requesting the use
+    of a proxy based on the address to which the subchannel is going
+    to connect.  If a proxy is to be used, the mapper will return the
+    proxy address (which the subchannel will connect to instead of the
+    address returned by the resolver) and add the additional state
+    described above to trigger the use of the HTTP CONNECT handshaker,
+    with the argument set to the original subchannel address.  Note that
+    in this case, the server name used in the HTTP CONNECT request will
+    be an IP address and port, not a hostname (i.e., we are not relying
+    on the proxy to do name resolution for us).
+
+Note that in case 1, because the client cannot know the set of server
+addresses, it is impossible to use the normal gRPC client-side per-call
+load balancing.  It *is* possible to do load balancing on a per-connection
+basis, but that may not spread out the load evenly if different clients
+impose a different amount of load.  (Note that even if we established
+multiple connections to the proxy, we would have no guarantee that each
+connection would go to a different backend server.)
+
+In contrast, in case 2, it *is* possible to do the normal gRPC client-side
+per-call load balancing, because the client does know the set of server
+addresses.  The fact that the connections to those individual addresses
+go through a TCP-level proxy does not interfere with the client opening
+a separate subchannel to each address.
+
+Case 3 is similar to case 1 in that the resolver cannot return the actual
+addresses of the servers.  However, we (ab)use the gRPC load-balancing
+design to work around this, so that it looks more like case 2.  In this
+case, there will be an external name service record for the server name
+that points to the IP address of the proxy and has the `is_balancer`
+bit set.  (Note: We have not yet designed how that bit will be encoded
+in DNS, but that will be the subject of a separate gRFC.)  The proxy
+mapper implementation will then have to detect two types of addresses:
+
+- When it sees the proxy address, it will set the HTTP CONNECT argument to
+  the original server name (which will cause the proxy to establish a
+  connection to one of the load balancers).
+- When it sees any internal address (i.e., the addresses returned by
+  the load balancer), it will replace it with the proxy address and set
+  the HTTP CONNECT argument to the internal address.
+
+Note that this will work only if the `grpclb` load balancing policy is
+in use; it will not work with client-side policies like `round_robin`.
+
+Also, note that in case 1, the client will not have access to the
```

Comment:
To expand on what Eric said, note that when this doc says that the proxy does the name resolution for us, that doesn't mean that the proxy actually provides the same result that we would get from the resolver; all it means is that we're sending the hostname to the proxy (as the argument to the CONNECT request) and depending on it do resolve that to one specific backend address when establishing the requested connection.

There's no reasonable way for the proxy to return the service config data.  Although it could conceivably send that information via a header in the CONNECT response, most proxies won't have any way of knowing to do that, because this type of proxy is not grpc-specific.  And even if it did, we don't have a reasonable way to make use of that information, because (as Eric points out) we would be obtaining that data in the subchannel connection code, which is lower in the stack than where we actually need the service config data.